### ISSUE TITLE
Add octal literals

### ISSUE THREAD

**Issue by nex3** [dart-lang/language#2708]
Related to https://github.com/dart-lang/language/issues/581, I'd also like to see octal literals. The lack of these actually caused a security issue for Dart Sass recently (see https://github.com/google/dart_cli_pkg/pull/109) because we incorrectly converted Unix permissions bits from octal to decimal and ended up creating world-writable executables under certain circumstances.

**Comment by lrhn** [comment]
Unix permissions is the only use for octal number that I'm aware of. 
It seems like something that could be fixed by using an abstraction, rather than adding a language feature with no other practical uses. Will probably be more readable than 0o754 to readers too.

Take:
```dart
const x = 1;
const w = 2;
const r = 4;
const rw = r | w;
const rx = r | x;
const wx = w | x;
const rwx = r | w | x;

int perm({int u = 0, int f = 0, int o = 0, int user = 0, int group = 0, int others = 0}) => 
   ((u | user) << 6) | ((g | group) << 3) | (o | others);
```
so you can just write `perm(u: rw, g: r, o: w)`, and not use (low-readability) number literals at all.

Use the longer names if you like that better. Add extra parameters for special bits. Or parse a string.

**Comment by mateusfccp** [comment]
Should we take advantage of https://github.com/dart-lang/language/issues/581#issuecomment-2259691898 and add octal literals? It seems @srawlins patch could be easily adapted to also solve this one.

**Comment by lrhn** [comment]
IMO, no thanks. I'd rather *not*.

Octal numbers have precisely one semi-common use (Unix file permissions). That's not enough of a reason to put them into the Dart language. There really is not enough use-cases for the feature to be worth the effort (and syntax real-estate), no matter how small that effort is.
(A library with 512 constants, `const p000 = 0;` through `const p777 = 0x1FF;`, will solve that one use-case.)

Dart doesn't have to support every base of literal, and the reason to have octal is so minusculy bigger than having quaternary that I don't think it should make a difference.
(Or if Dart _should_ support all radixes, then have one syntax parameterized with the base, say `8r775`, with `r` meaning "radix", since `b` for "base" is taken by binary.).

**Comment by nex3** [comment]
To my mind, the most compelling difference between octal and quaternary is that octal is widely supported by programming languages using a consistent, well-known syntax and quaternary is not. Realistically, it's not giving up "syntax real-estate" to support it because if you defined `0o755` to mean anything _other than_ 493, you'd be direly undermining user expectations and even potentially posing a security risk by making it easy to write code that looked like it was permissions-correct but in fact was not.

Dart isn't an abstract spec sitting in the realm of platonic forms—it's a programming language in the real world, used by programmers who use many other languages and have expectations accordingly. This doesn't mean it always has to do everything the same, but when there's a clear expectation that's very inexpensive to meet and provides real value, why not meet it?

**Comment by lrhn** [comment]
If no other language had octal literals, would we add them?
Probably not. Just too obscure a use-case. 

But other languages do have octal literals. 
Some use `045` as syntax. That's a horrible design mistake that should not be repeated. 
Others use `0o45` which makes sense by itself. Not great for reading, and particularly bad if it also allows a capital `O`, `0O45` is just not readable. I wouldn't allow that, lower case only if we ever did it. 

We don't have to do things just because other languages do. 
If other languages do something, and lots of uses use it and expect it, that's another thing. As yours say, if there is a clear expectation, we probably should meet it. 

But there isn't. Octal literals are not table stakes, there esoteric features. 
Of the languages you mention above: JavaScript, Java, Go, Python, Ruby, Rust, C, and Swift, hour many can you actually remember the syntax for octals in?

I remember one, JavaScript. I'd guess that Java and C have the same syntax, but it is a guess. I wrote enterprises Java apps for years and never had to use an octal literal. I'd wager that the majority of Java developers are unaware that octal literals exist.

It's an obscured feature that I actually think adds more complexity to a language than it's worth. Even if it's not a lot of complexity.

**Comment by munificent** [comment]
> Of the languages you mention above: JavaScript, Java, Go, Python, Ruby, Rust, C, and Swift, hour many can you actually remember the syntax for octals in? Or that they had octal numbers at all, without looking it up?

I was curious:

* JavaScript supported both old-style `0123` octal literals and `0o123` ones. The former are [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal_literal).

* Java [uses the crappy `0123` syntax](https://softwareengineering.stackexchange.com/questions/221797/reasoning-behind-the-syntax-of-octal-notation-in-java).

* Go [supports both old-style `0123` and `0o123` ones](https://go.dev/doc/go1.13#language). The latter was added later.

* Python supported old-style but [moved to `0o123` in Python 2.6](https://peps.python.org/pep-3127/).

* Ruby [supports both `0123` and `0o123`](https://docs.ruby-lang.org/en/2.0.0/syntax/literals_rdoc.html). (It also supports `0d123` if you want to be explicit that you're writing in decimal, which is cute.)

* Rust [only supports `0o123` style](https://doc.rust-lang.org/rust-by-example/primitives/literals.html).

* C of course invented the old cursed style.

* Swift [uses `0o123` octal literals](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/lexicalstructure/).

So, aside from C and Java, every listed language supports `0o123`-style octal literals. A few support the old style ones too.

**Comment by nex3** [comment]
The last case I'll make before I leave this up to the language team: although it's true that octal literals have narrow use-cases relative to binary and hexadecimal literals, the use-cases they do have are specifically security-related. The cost of getting a manual conversion to octal wrong is substantially higher than the cost of an average bug. I think this, as well as the wide availability of the `0o` style that @munificent demonstrated, tips the balance into making this a worthwhile addition.

**Comment by lrhn** [comment]
If we can help people not make mistakes, we should. 
A "You look like you want to write an octal literal" message could help, either for anything of the form `0123`, decimal with leading zero and no 8 or 9 digit, but that might be too harsh, or instead targeting known "permission" parameters, or anything annotated with a `@userMightUseOctal` marker, or whatever works.

I don't think actually adding octal numerals is the best solution to any problem. Especially if nobody will ever use more than the three digits.

So far I've suggested symbolic constants, or even a library with 512 constants from `o000` to `o777`.
Now we have an event better option: binary numerals. Rather than `0o755` you can write `0b111_101_101`.
I think it's easier to read too!
If C had had binary numerals, maybe that would have been how you wrote permissions all along.

**Comment by tatumizer** [comment]
Fun fact.
Here's my dialog with gemini AI.
Quote:
Me: Example of octal number literal in dart programming language
Gemini:
```dart
void main() {
  int octalNumber = 0o177; // 0o prefix indicates an octal literal
  print("Octal number: $octalNumber"); // Output: Octal number: 127
}
```
Explanation:

0o prefix: This prefix explicitly signifies that the following digits represent an octal number.
Digits: Octal numbers use digits from 0 to 7.
Output: The code will print the decimal equivalent of the octal number 0o177, which is 127.
Key Points:

Octal literals are a concise way to represent numbers in base-8 within your Dart code.
The 0o prefix is crucial to distinguish octal numbers from decimal or other number systems.
I hope this example helps!
End Quote

:smile:

**Comment by lrhn** [comment]
Generative AIs be hallucinating. Must be Tuesday.

Adding `0o123` as an octal literal won't solve the problem of someone writing `0755` and getting a decimal interpretation. Solving that requires disallowing, or at least warning about, leading zeros in decimal integer numerals. If we do that, then we can redirect you to any solution, it doesn't have to be octal literals.

**Comment by tatumizer** [comment]
True. I am not aware of any legitimate use case for decimal literals with the leading zero, so the warning seems well-justified.
The problem is that `0o177` literals seem to be widespread (that's the reason for Gemini's hallucinations). It's a philosophical issue: the programming idioms keep changing, and `0o177` looks like a current idiom (replacing legacy 0177). It might be easier to support it than to explain to everyone why you don't want to support it. (These `0o` literals won't interfere with existing or future dart features).

**Comment by lrhn** [comment]
A quick search found a few occurrences of leading zeros:
* Time/data which is traditionally two-digit, like `DateTime(2024, 03, 28, 10, 00, 00)`
* Tabular data, like this [base64 table](https://github.com/dart-lang/sdk/blob/main/sdk/lib/convert/base64.dart#L596)
* Bugs ([I think](https://github.com/halildurmus/win32/blob/96a7b975ed184d10512f434db50b493beca4485c/packages/win32/lib/src/constants.dart#L10030))
* Some special-formatted composite numbers, like a date being `19840101` and then having 'minDecimalData` be `00000101`.

The *vast* majority are `DateTime` arguments.

**Comment by tatumizer** [comment]
Only DateTime seems legit (though hardcoded dates must be uncommon other than it tests). Maybe that's the use case for `0d`?
Also (quoting the docs):
"While the 0d prefix is rarely used (because decimal is the default in Ruby), it is supported and can be helpful for clarity when working with mixed numeric bases".

**Comment by tatumizer** [comment]
@lumpidu: please refer to this comment for the context:
https://github.com/dart-lang/language/issues/2708#issuecomment-2284909963

**Comment by munificent** [comment]
> Only DateTime seems legit (though hardcoded dates must be uncommon other than it tests). Maybe that's the use case for `0d`?

I would bet money that the number of people who expect to be able to write `03` for the third day in a month is larger than the set of people who expect to write `0666` to set file permissions for a POSIX machine.

**Comment by lrhn** [comment]
There are only 512 three-digit octal literals. Well within the range of just having a constant for each: https://gist.github.com/lrhn/9cf3249827503c1632cbb292b753ebdf

We could have the analyzer comment on any decimal literal with (a single?) leading zero, no `8` or `9` digits and more than one significant digit. Not a warning, just an "unnecessary leading digit" hint.
No warnings for `099` (contains `9`), `01` (same value if octal), maybe `0011` (two leading zeros, not necessary for an old-style octal, but `0055` could be intended as a permission).

I'd still get dinged on `DateTime(0100, 01, 01)`, written that way to align with other dates in the same table, but I'll only use literal pre-1970 dates in tests anyway, so the problem is constrained.

Might get a comment on a table like
```dart
// Not a prime.
const _np = 0;
// First 128 numbers, if they are prime, zero if not.
const primes = [ // Nobreak.
  _np, _np, 002, 003, _np, 005, _np, 007, _np, _np, _np, 011, _np, 013, _np, _np, _np, 017, _np, 019,
  _np, _np, _np, 023, _np, _np, _np, _np, _np, 029, _np, 031, _np, _np, _np, _np, _np, 037, _np, _np,
  _np, 041, _np, 043, _np, _np, _np, 047, _np, _np, _np, _np, _np, 053, _np, _np, _np, _np, _np, 059,
  _np, 061, _np, _np, _np, _np, _np, 067, _np, _np, _np, 071, _np, 073, _np, _np, _np, _np, _np, 079,
  _np, _np, _np, 083, _np, _np, _np, _np, _np, 089, _np, _np, _np, _np, _np, _np, _np, 097, _np, _np,
  _np, 101, _np, 103, _np, _np, _np, 107, _np, 109, _np, _np, _np, 113, _np, _np, _np, _np, _np, _np,
  _np, _np, _np, _np, _np, _np, _np, 127];
```
The analyzer could try to be a little clever with collections that contain multiple number literals, and only
say something if they're all "possibly intended as octals".

**Comment by tatumizer** [comment]
But such a (rare) program can use `// ignore: ` comment, no?

**Comment by lrhn** [comment]
Absolutely. (And if I had `// ignore` for a region, not the entire file, it would be even better.)

My main worry is whether it introduces a buch of warnings in existing code that authors have to go back and fix up.
(But if lints can be language versioned, we could just say that it doesn't trigger until you upgrade to some language version.)
